tests/test_json.py

```diff
@@ -0,0 +1,58 @@
+import datetime
+import json
+from decimal import Decimal
+from enum import Enum
+from uuid import UUID
+
+import pytest
+
+from pydantic import BaseModel, create_model
+from pydantic.json import pydantic_encoder
+
+
+class MyEnum(Enum):
+    foo = 'bar'
+    snap = 'crackle'
+
+
+@pytest.mark.parametrize('input,output', [
+    (UUID('ebcdab58-6eb8-46fb-a190-d07a33e9eac8'), '"ebcdab58-6eb8-46fb-a190-d07a33e9eac8"'),
+    (datetime.datetime(2032, 1, 1, 1, 1), '"2032-01-01T01:01:00"'),
+    (datetime.datetime(2032, 1, 1, 1, 1, tzinfo=datetime.timezone.utc), '"2032-01-01T01:01:00+00:00"'),
+    (datetime.datetime(2032, 1, 1), '"2032-01-01T00:00:00"'),
+    (datetime.time(12, 34, 56), '"12:34:56"'),
+    ({1, 2, 3}, '[1, 2, 3]'),
+    (frozenset([1, 2, 3]), '[1, 2, 3]'),
+    ((v for v in range(4)), '[0, 1, 2, 3]'),
+    (b'this is bytes', '"this is bytes"'),
+    (Decimal('12.34'), '12.34'),
+    (create_model('BarModel', a='b', c='d')(), '{"a": "b", "c": "d"}'),
+    (MyEnum.foo, '"bar"')
+])
+def test_encoding(input, output):
+    assert json.dumps(input, default=pydantic_encoder) == output
+
+
+def test_model_encoding():
+    class ModelA(BaseModel):
+        x: int
+        y: str
+
+    class Model(BaseModel):
+        a: float
+        b: bytes
+        c: Decimal
+        d: ModelA
+
+    m = Model(a=10.2, b='foobar', c=10.2, d={'x': 123, 'y': '123'})
+    assert m.dict() == {'a': 10.2, 'b': b'foobar', 'c': Decimal('10.2'), 'd': {'x': 123, 'y': '123'}}
+    assert m.json() == '{"a": 10.2, "b": "foobar", "c": 10.2, "d": {"x": 123, "y": "123"}}'
+    assert m.json(exclude={'b'}) == '{"a": 10.2, "c": 10.2, "d": {"x": 123, "y": "123"}}'
```

Comment:
While its nice to have this call-site instance-level exclude flexibility, I think most use-cases will be based on standard `exclude`s at the definition level (aka. Class). Is it already possible to do that?

Reply:
no, for `dict` and `copy` exclude and include are provided here. I'd like to keep it the same.

Reply:
Just thinking that _in addition_ to this there should be a way to specify this at the model definition level. In Go for instance they use annotations feature on structs to define how it maps to JSON when serialized.

Reply:
And actually this is effectively a new feature so should be a new issue/pr etc. anyways... carry on 😅

Reply:
interesting, but as you say, definitely separate from this.